File: csharp/src/Callisto.Sdk/CallistoClient.cs
using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Threading.Tasks;
using Callisto.Sdk.Http;
using Callisto.Sdk.Reporting;
using Callisto.Sdk.Resources;

namespace Callisto.Sdk;

/// <summary>
/// The Callisto API client. Construct with credentials (or let them resolve from environment
/// variables) and access resources via properties: <see cref="Balance"/>, <see cref="Sms"/>,
/// <see cref="Otp"/>, <see cref="WhatsApp"/>, <see cref="Notify"/>.
/// </summary>
public sealed class CallistoClient : IDisposable
{
    private readonly Transport _transport;
    private readonly ErrorReporter _reporter;
    private readonly bool _captureUnhandled;
    private UnhandledExceptionEventHandler? _domainHandler;
    private EventHandler<UnobservedTaskExceptionEventArgs>? _taskHandler;

    /// <summary>Account balance operations.</summary>
    public BalanceResource Balance { get; }

    /// <summary>SMS operations.</summary>
    public SmsResource Sms { get; }

    /// <summary>One-time-password operations.</summary>
    public OtpResource Otp { get; }

    /// <summary>WhatsApp operations.</summary>
    public WhatsAppResource WhatsApp { get; }

    /// <summary>Multi-channel notification operations.</summary>
    public NotifyResource Notify { get; }

    /// <summary>
    /// The built-in error reporter. Disabled (no-op) unless an error DSN is configured. The three
    /// client methods (<see cref="CaptureException"/>, <see cref="CaptureMessage"/>,
    /// <see cref="SetUser"/>) are the supported surface; this is exposed for advanced use.
    /// </summary>
    public ErrorReporter ErrorReporter => _reporter;

    /// <summary>
    /// Creates a client. Any of <paramref name="clientId"/>, <paramref name="apiKey"/>, and
    /// <paramref name="baseUrl"/> fall back to <c>CALLISTO_CLIENT_ID</c>,
    /// <c>CALLISTO_API_KEY</c>, and <c>CALLISTO_BASE_URL</c> respectively. Throws
    /// <see cref="ArgumentException"/> if the client ID or API key cannot be resolved.
    /// </summary>
    /// <param name="clientId">Your Callisto client ID.</param>
    /// <param name="apiKey">Your Callisto API key.</param>
    /// <param name="baseUrl">API base URL. Defaults to <see cref="Config.DefaultBaseUrl"/>.</param>
    /// <param name="timeout">Request timeout. Defaults to 30 seconds.</param>
    /// <param name="httpClient">Optional pre-configured <see cref="HttpClient"/> to inject.</param>
    /// <param name="handler">
    /// Optional <see cref="HttpMessageHandler"/> to build the internal client on (useful for
    /// tests). Ignored when <paramref name="httpClient"/> is supplied.
    /// </param>
    /// <param name="errorDsn">
    /// Error-reporting ingest DSN. Falls back to <c>CALLISTO_ERROR_DSN</c>. Absent → reporting
    /// disabled.
    /// </param>
    /// <param name="captureUnhandled">
    /// Install the global unhandled-exception handler. Falls back to
    /// <c>CALLISTO_CAPTURE_UNHANDLED</c>. Defaults to <c>false</c>.
    /// </param>
    /// <param name="environment">Optional environment tag. Falls back to <c>CALLISTO_ENVIRONMENT</c>.</param>
    /// <param name="errorSender">Optional injectable sender for the reporter (useful for tests).</param>
    public CallistoClient(
        string? clientId = null,
        string? apiKey = null,
        string? baseUrl = null,
        TimeSpan? timeout = null,
        HttpClient? httpClient = null,
        HttpMessageHandler? handler = null,
        string? errorDsn = null,
        bool? captureUnhandled = null,
        string? environment = null,
        IErrorSender? errorSender = null)
    {
        var config = Config.Resolve(clientId, apiKey, baseUrl, timeout, errorDsn, captureUnhandled, environment);

        _reporter = new ErrorReporter(config.ErrorDsn, config.Environment, errorSender);
        _captureUnhandled = config.CaptureUnhandled;

        _transport = new Transport(config, httpClient, handler, _reporter);
        Balance = new BalanceResource(_transport);
        Sms = new SmsResource(_transport);
        Otp = new OtpResource(_transport);
        WhatsApp = new WhatsAppResource(_transport);
        Notify = new NotifyResource(_transport);

        if (_captureUnhandled && _reporter.Enabled)
        {
            InstallGlobalHandler();
        }
    }

    /// <summary>Captures an exception via the error reporter. No-op when reporting is disabled.</summary>
    public void CaptureException(
        Exception ex,
        string level = "error",
        IDictionary<string, object?>? extra = null)
        => _reporter.CaptureException(ex, level, extra);

    /// <summary>Captures a plain message via the error reporter. No-op when reporting is disabled.</summary>
    public void CaptureMessage(
        string message,
        string level = "info",
        IDictionary<string, object?>? extra = null)
        => _reporter.CaptureMessage(message, level, extra);

    /// <summary>Sets/clears the user context attached to subsequent captured events.</summary>
    public void SetUser(IDictionary<string, object?>? user) => _reporter.SetUser(user);

    private void InstallGlobalHandler()
    {
        _domainHandler = (_, args) =>
        {
            if (args.ExceptionObject is Exception ex)
            {
                _reporter.CaptureException(ex, "fatal");
                _reporter.Flush();
            }
        };
        AppDomain.CurrentDomain.UnhandledException += _domainHandler;

        _taskHandler = (_, args) =>
        {
            // Capture but do not mark observed — preserve default behavior.
            _reporter.CaptureException(args.Exception, "fatal");
        };
        TaskScheduler.UnobservedTaskException += _taskHandler;
    }

    /// <summary>Releases the underlying HTTP resources and flushes the error reporter.</summary>
    public void Dispose()
    {
        if (_domainHandler is not null)
        {
            AppDomain.CurrentDomain.UnhandledException -= _domainHandler;
            _domainHandler = null;
        }

        if (_taskHandler is not null)
        {
            TaskScheduler.UnobservedTaskException -= _taskHandler;
            _taskHandler = null;
        }

        _reporter.Dispose();
        _transport.Dispose();
    }
}

File: java/src/main/java/com/callisto/sdk/CallistoClient.java
package com.callisto.sdk;

import com.callisto.sdk.http.Transport;
import com.callisto.sdk.reporting.ErrorReporter;
import com.callisto.sdk.reporting.ErrorSender;
import com.callisto.sdk.resources.Balance;
import com.callisto.sdk.resources.Notify;
import com.callisto.sdk.resources.Otp;
import com.callisto.sdk.resources.Sms;
import com.callisto.sdk.resources.WhatsApp;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.net.http.HttpClient;
import java.time.Duration;
import java.util.Map;

/**
 * Entry point for the Callisto SDK.
 *
 * <p>Authentication is HTTP Basic ({@code clientId} / {@code apiKey}) and is applied
 * automatically to every request. Implements {@link AutoCloseable}.
 *
 * <pre>{@code
 * try (CallistoClient client = new CallistoClient("your-client-id", "your-api-key")) {
 *     Balance balance = client.balance().get();
 * }
 * }</pre>
 */
public class CallistoClient implements AutoCloseable {

    private final Transport transport;
    private final ErrorReporter errorReporter;
    private final Balance balance;
    private final Sms sms;
    private final Otp otp;
    private final WhatsApp whatsapp;
    private final Notify notify;

    /**
     * Creates a client using the given credentials (or {@code null} to fall back to the
     * {@code CALLISTO_CLIENT_ID} / {@code CALLISTO_API_KEY} environment variables), the default
     * base URL, and a 30s timeout.
     */
    public CallistoClient(String clientId, String apiKey) {
        this(clientId, apiKey, null, null);
    }

    /**
     * Creates a client.
     *
     * @param clientId Callisto client ID (falls back to env {@code CALLISTO_CLIENT_ID}).
     * @param apiKey   Callisto API key (falls back to env {@code CALLISTO_API_KEY}).
     * @param baseUrl  optional API base URL (falls back to env {@code CALLISTO_BASE_URL},
     *                 then the default). Trailing slashes are trimmed.
     * @param timeout  optional request timeout (defaults to 30s).
     */
    public CallistoClient(String clientId, String apiKey, String baseUrl, Duration timeout) {
        this(Config.resolve(clientId, apiKey, baseUrl, timeout), null, null, null);
    }

    /**
     * Creates a client with an injectable {@link HttpClient} and {@link ObjectMapper}
     * (advanced use, e.g. testing or custom transport configuration). Pass {@code null} to
     * use the defaults.
     */
    public CallistoClient(String clientId, String apiKey, String baseUrl, Duration timeout,
                          HttpClient httpClient, ObjectMapper mapper) {
        this(Config.resolve(clientId, apiKey, baseUrl, timeout), httpClient, mapper, null);
    }

    /**
     * Creates a client with full error-reporting configuration and an injectable
     * {@link ErrorSender} (advanced use / testing). Pass {@code null} for any optional argument
     * to fall back to defaults / environment variables.
     */
    public CallistoClient(String clientId, String apiKey, String baseUrl, Duration timeout,
                          HttpClient httpClient, ObjectMapper mapper,
                          String errorDsn, Boolean captureUnhandled, String environment,
                          ErrorSender errorSender) {
        this(Config.resolve(clientId, apiKey, baseUrl, timeout, errorDsn, captureUnhandled, environment),
                httpClient, mapper, errorSender);
    }

    private CallistoClient(Config config, HttpClient httpClient, ObjectMapper mapper,
                           ErrorSender errorSender) {
        ObjectMapper resolvedMapper = mapper != null ? mapper : new ObjectMapper();
        this.transport = httpClient != null
                ? new Transport(config, httpClient, resolvedMapper)
                : new Transport(config);

        this.errorReporter = new ErrorReporter(
                config.getErrorDsn(), config.getEnvironment(), errorSender, resolvedMapper);
        if (errorReporter.isEnabled()) {
            transport.setReporter(errorReporter);
        }

        this.balance = new Balance(transport);
        this.sms = new Sms(transport);
        this.otp = new Otp(transport);
        this.whatsapp = new WhatsApp(transport);
        this.notify = new Notify(transport);

        if (config.isCaptureUnhandled()) {
            errorReporter.installUncaughtHandler();
        }
    }

    public Balance balance() {
        return balance;
    }

    public Sms sms() {
        return sms;
    }

    public Otp otp() {
        return otp;
    }

    public WhatsApp whatsapp() {
        return whatsapp;
    }

    /**
     * The notify resource.
     *
     * <p>Named {@code notify()} would collide with the {@code final} {@link Object#notify()};
     * this accessor is therefore named {@code notifications()}.
     */
    public Notify notifications() {
        return notify;
    }

    /**
     * The error reporter (advanced use). Reachable for direct configuration; the three
     * {@code capture*} / {@code setUser} client methods are the supported surface. Always
     * non-null; a no-op when no DSN is configured.
     */
    public ErrorReporter errorReporter() {
        return errorReporter;
    }

    /**
     * Reports an exception to the error-tracking endpoint at {@code level = error}.
     * No-op when no DSN is configured. Never throws.
     */
    public void captureException(Throwable t) {
        errorReporter.captureException(t);
    }

    /** Reports an exception at the given level. No-op without a DSN. Never throws. */
    public void captureException(Throwable t, String level, Map<String, Object> extra) {
        errorReporter.captureException(t, level, extra);
    }

    /**
     * Reports a plain message to the error-tracking endpoint at {@code level = info}.
     * No-op when no DSN is configured. Never throws.
     */
    public void captureMessage(String message) {
        errorReporter.captureMessage(message);
    }

    /** Reports a plain message at the given level. No-op without a DSN. Never throws. */
    public void captureMessage(String message, String level, Map<String, Object> extra) {
        errorReporter.captureMessage(message, level, extra);
    }

    /** Sets or clears the {@code user} context attached to subsequent reported events. */
    public void setUser(Map<String, Object> user) {
        errorReporter.setUser(user);
    }

    @Override
    public void close() {
        errorReporter.close();
        transport.close();
    }
}
